### PR TITLE
added wiki feature

### PR DESCRIPTION
**Pull request by priyansh1161**
fixed permission to 644 and changed somethings in wiki

### CODE REVIEW

**Comment by 526avijitgupta** [comment]
@priyansh1161 Fix the file permissions of chatbot.js to 644 instead of 755, so that the diff is correctly parsed!

File: services/wikiService.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
+var request = require('request');
+
+module.exports = function (cb,username,input) {
+    'use strict';
+    let apiOptions = {
+        url :`https://en.wikipedia.org/w/api.php?format=json&action=query&generator=search&gsrnamespace=0&gsrlimit=4&prop=pageimages|extracts&pilimit=max&exintro&explaintext&exsentences=1&exlimit=max&gsrsearch=${input}`,
```
Comment (yashmehrotra) [inline services/wikiService.js]: @priyansh1161 Move this url to a different file.
@526avijitgupta Please suggest the best way of storing different urls for various API Services ?

File: services/wikiService.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+        method : 'GET',
+        json : {}
```
Comment (yashmehrotra) [inline services/wikiService.js]: @priyansh1161 Just remove the `json: {}`, looks cleaner

File: services/wikiService.js [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    };
+    request(apiOptions,(err,response,data)=>{
+       if(!err && response.statusCode==200){
+           let result = '',dataArr=[];
```
Comment (yashmehrotra) [inline services/wikiService.js]: @priyansh1161 Space after comma

File: services/wikiService.js [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+           if(data.query){
+               for(var key in data.query.pages){
+                   dataArr[data.query.pages[key].index-1 ]=data.query.pages[key]; //counting sort indexed at 0 .
+               }
+           }
+           dataArr.forEach((curr,index)=>{
+               result+=`\n ## ${index+1}[${curr.title}](https://en.wikipedia.org/?curid='${curr.pageId})\n ${curr.extract}\n`;
```
Comment (yashmehrotra) [inline services/wikiService.js]: @priyansh1161 use `###`, the `##` is a bit too big

File: services/wikiService.js (hunk 3 above)
Comment (526avijitgupta) [reply services/wikiService.js]: Space before and after operators as well (=, ==, => etc..)

File: services/wikiService.js (hunk 4 above)
Comment (526avijitgupta) [reply services/wikiService.js]: Spaces after {, } as well.

Comment (yashmehrotra) [reply services/wikiService.js]: @priyansh1161 Also space b/w `${index+1}` ad `[${curr.titl..`

File: services/wikiService.js [hunk 5]
```diff
@@ -0,0 +1,25 @@
+var request = require('request');
+
+module.exports = function (cb,username,input) {
+    'use strict';
```
Comment (526avijitgupta) [inline services/wikiService.js]: Move this to the top of this file!

File: services/wikiService.js [hunk 6]
```diff
@@ -0,0 +1,25 @@
+var request = require('request');
+
+module.exports = function (cb,username,input) {
```
Comment (526avijitgupta) [inline services/wikiService.js]: Rename cb to something more descriptive and clear, spaces around `,`

File: services/wikiService.js (hunk 1 above)
Comment (526avijitgupta) [reply services/wikiService.js]: @yashmehrotra I'm implementing a better method as a part of my own PR.
@priyansh1161 For now, please declare a const and use it for the url.

File: services/wikiService.js [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+           let result = '',dataArr=[];
+           if(data.query){
+               for(var key in data.query.pages){
+                   dataArr[data.query.pages[key].index-1 ]=data.query.pages[key]; //counting sort indexed at 0 .
+               }
+           }
+           dataArr.forEach((curr,index)=>{
+               result+=`\n ## ${index+1}[${curr.title}](https://en.wikipedia.org/?curid='${curr.pageId})\n ${curr.extract}\n`;
+           });
+           console.log(result);
```
Comment (526avijitgupta) [inline services/wikiService.js]: Remove this.

File: services/wikiService.js [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+           cb(result,username);
+    }
+    });
+};
```
Comment (526avijitgupta) [inline services/wikiService.js]: Add newline after end of file.

File: services/wikiService.js [hunk 9]
```diff
@@ -0,0 +1,27 @@
+var request = require('request');
+var URL =  'https://en.wikipedia.org/w/api.php?format=json&action=query&generator=search&gsrnamespace=0&gsrlimit=4&prop=pageimages|extracts&pilimit=max&exintro&explaintext&exsentences=1&exlimit=max&gsrsearch=';
```
Comment (526avijitgupta) [inline services/wikiService.js]: Rename to WIKI_PREFIX_URL. Also, leave a newline after this.

File: services/wikiService.js [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+module.exports = function (cb,username,input) {
+    'use strict';
+    let apiOptions = {
+        url :`${URL}${input}`,
+        method : 'GET',
+        json : {} //it is required
```
Comment (526avijitgupta) [inline services/wikiService.js]: I don't think the comment is needed here. Anyone who sees this code in the future might get confused on reading this.

**Comment by yashmehrotra** [comment]
@526avijitgupta Can you please look into what is happening to chatbot.js ?

**Comment by 526avijitgupta** [comment]
@yashmehrotra Will do. Once @priyansh1161 fixes the file permissions as I mentioned in my previous comment.

**Comment by yashmehrotra** [comment]
@526avijitgupta Its not showing a diff in permission level

**Comment by 526avijitgupta** [comment]
@yashmehrotra sorry, what do you mean ?

**Comment by priyansh1161** [comment]
i have fixed file permissions already using chmod 644 chatbot.js

**Comment by 526avijitgupta** [comment]
@priyansh1161 The diff is parsing incorrectly. Most probably due to line endings. Check that `git config core.autocrlf` is not true. If it is, then change it to false and re-commit chatbot.js thereafter.

**Comment by priyansh1161** [comment]
@526avijitgupta what went wrong this time ?

**Comment by yashmehrotra** [comment]
@priyansh1161 Ok, just 2 more things to do:
1. Lint your JS Code.
2. Make the whole thing into a single commit

**Comment by priyansh1161** [comment]
@526avijitgupta @yashmehrotra what ESlint configure file are you guys using ? i think my config is different 

``` javascript
{
    "env": {
        "node": true,
        "commonjs": true,
        "es6": true
    },
    "extends": "eslint:recommended",
    "rules": {
        "indent": [
            "error",
            4
        ],
        "linebreak-style": [
            "error",
            "unix"
        ],
        "quotes": [
            "error",
            "single"
        ],
        "semi": [
            "error",
            "always"
        ]
    }
}
```

**Comment by 526avijitgupta** [comment]
@priyansh1161 There isn't a common eslint config yet (I'll be adding one after the existing opened PRs are merged). But it's evidently clear that the indentation used across the code is 2 - spaced.
Also, you can skip the 2nd step as mentioned by Yash - no need to squash all commits into one.

/cc @yashmehrotra -- squash and merge is available on GitHub.

**Comment by 526avijitgupta** [comment]
@priyansh1161 - Another thing, just noticed that this PR has been generated out of the master branch from your fork. I guess we'll be merging it this time (once the remaining comments are addressed and indentation is fixed), but please make a habit to create a feature branch out of master and generating a pull request from it.

**Comment by priyansh1161** [comment]
@526avijitgupta will remember from next time. 
P.S i have addressed all comments in my last PR

File: chatbot.js [hunk 11]
```diff
@@ -69,7 +70,6 @@ const messageHandler = (msg) => {
     reply_to_user(msg.model.fromUser, msg.model.text);
   }
 };
-
 function _getStartsWith(parsedMessage) {
```
Comment (526avijitgupta) [inline chatbot.js:74]: Leave line above this function.

File: package.json [hunk 12]
```diff
@@ -11,5 +11,8 @@
   "dependencies": {
     "faye": "~1.0.1",
     "request": "^2.74.0"
+  },
+  "devDependencies": {
+    "eslint": "^3.4.0"
```
Comment (526avijitgupta) [inline package.json]: I don't think this is used anywhere. Please reset the changes here.

File: chatbot.js [hunk 13]
```diff
@@ -126,6 +126,10 @@ function reply_to_user(user, message) {
         send(body);
       });
     }
+    if(startsWithString === BOT_ACTIONS.WIKI){
```
Comment (526avijitgupta) [inline chatbot.js]: replace by `if (startsWithString === BOT_ACTIONS.WIKI) {` - note the spaces.

File: chatbot.js [hunk 14]
```diff
@@ -126,6 +126,10 @@ function reply_to_user(user, message) {
...
+      var requestedData = parsedMessage.slice(5,parsedMessage.length);
+      wiki(send,username,requestedData);
```
Comment (526avijitgupta) [inline chatbot.js]: spaces after comma on both lines.

File: services/wikiService.js [hunk 15]
```diff
@@ -0,0 +1,29 @@
+var request = require('request');
+var WIKI_PREFIX_URL =  'https://en.wikipedia.org/w/api.php?format=json&action=query&generator=search&gsrnamespace=0&gsrlimit=4&prop=pageimages|extracts&pilimit=max&exintro&explaintext&exsentences=1&exlimit=max&gsrsearch=';
+
+module.exports = function (callback,username,input){
+  'use strict';
```
Comment (526avijitgupta) [inline services/wikiService.js]: Move this to the beginning of this file.

File: services/wikiService.js [hunk 16]
```diff
@@ -0,0 +1,29 @@
+var request = require('request');
+var WIKI_PREFIX_URL =  'https://en.wikipedia.org/w/api.php?format=json&action=query&generator=search&gsrnamespace=0&gsrlimit=4&prop=pageimages|extracts&pilimit=max&exintro&explaintext&exsentences=1&exlimit=max&gsrsearch=';
```
Comment (526avijitgupta) [inline services/wikiService.js]: Use `const` instead of `var` on both lines.

File: services/wikiService.js [hunk 17]
```diff
@@ -0,0 +1,29 @@
...
+
+module.exports = function (callback,username,input){
```
Comment (526avijitgupta) [inline services/wikiService.js]: replace by `module.exports = (callback, username, input) => {`

File: services/wikiService.js [hunk 18]
```diff
@@ -0,0 +1,29 @@
...
+  'use strict';
+  let apiOptions = {
+    url :`${WIKI_PREFIX_URL}${input}`,
+    method : 'GET',
+    json : {}
+  };
+  request(apiOptions,(err,response,data) => {
+    if(!err && response.statusCode == 200){
+      let result = '', dataArr=[];
```
Comment (526avijitgupta) [inline services/wikiService.js]: `dataArr = []`

File: services/wikiService.js [hunk 19]
```diff
@@ -0,0 +1,29 @@
...
+      if(data.query){
```
Comment (526avijitgupta) [inline services/wikiService.js]: `if (data.query) {`

File: services/wikiService.js [hunk 20]
```diff
@@ -0,0 +1,29 @@
...
+        for(var key in data.query.pages){
```
Comment (526avijitgupta) [inline services/wikiService.js]: `for (let key in data.query.pages) {`

File: services/wikiService.js [hunk 21]
```diff
@@ -0,0 +1,29 @@
...
+          if(data.query.pages.hasOwnProperty(key))
```
Comment (526avijitgupta) [inline services/wikiService.js]: space before `(`

File: services/wikiService.js [hunk 22]
```diff
@@ -0,0 +1,29 @@
...
+            dataArr[data.query.pages[key].index-1 ]=data.query.pages[key]; //counting sort .
```
Comment (526avijitgupta) [inline services/wikiService.js]: `dataArr[data.query.pages[key].index - 1] = data.query.pages[key]; //counting sort`

File: services/wikiService.js [hunk 23]
```diff
@@ -0,0 +1,29 @@
...
+        }
+      }
+      dataArr.forEach((curr,index)=>{
```
Comment (526avijitgupta) [inline services/wikiService.js]: `dataArr.forEach((curr, index) => {`
P.S. nice work using forEach.

File: services/wikiService.js [hunk 24]
```diff
@@ -0,0 +1,29 @@
...
+        result+=`\n ### ${ index+1 }[${ curr.title }](https://en.wikipedia.org/?curid=${ curr.pageid })\n ${ curr.extract }\n`;
```
Comment (526avijitgupta) [inline services/wikiService.js]: result += `\n ### ${index+1}[${curr.title}](https://en.wikipedia.org/?curid=${curr.pageid})\n ${curr.extract}\n`;

File: services/wikiService.js [hunk 25]
```diff
@@ -0,0 +1,29 @@
...
+      });
+      callback(result,username);
```
Comment (526avijitgupta) [inline services/wikiService.js]: space after comma

File: services/wikiService.js [hunk 26]
```diff
@@ -0,0 +1,29 @@
...
+    }
+    else {
+      callback(err ?err.message :'unable to find',username);
```
Comment (526avijitgupta) [inline services/wikiService.js]: `callback(err ? err.message : 'unable to find', username);`
Also, replace `unable to find` with something a bit more descriptive.

**Comment by 526avijitgupta** [comment]
@priyansh1161 Great work 😄 . The comments I left are mostly styling related. The business logic is sound.

File: services/wikiService.js [hunk 27]
```diff
@@ -0,0 +1,29 @@
...
+
+module.exports = function (callback,username,input){
+  'use strict';
+  let apiOptions = {
+    url :`${WIKI_PREFIX_URL}${input}`,
+    method : 'GET',
+    json : {}
+  };
+  request(apiOptions,(err,response,data) => {
+    if(!err && response.statusCode == 200){
```
Comment (526avijitgupta) [inline services/wikiService.js]: `if (!err && response.statusCode === 200) {`

File: services/wikiService.js [hunk 28]
```diff
@@ -0,0 +1,29 @@
...
+var WIKI_PREFIX_URL =  'https://en.wikipedia.org/w/api.php?format=json&action=query&generator=search&gsrnamespace=0&gsrlimit=4&prop=pageimages|extracts&pilimit=max&exintro&explaintext&exsentences=1&exlimit=max&gsrsearch=';
+
+module.exports = function (callback,username,input){
+  'use strict';
+  let apiOptions = {
+    url :`${WIKI_PREFIX_URL}${input}`,
+    method : 'GET',
+    json : {}
+  };
+  request(apiOptions,(err,response,data) => {
```
Comment (526avijitgupta) [inline services/wikiService.js]: spaces after every comma.

File: services/wikiService.js [hunk 29]
```diff
@@ -0,0 +1,29 @@
+var request = require('request');
+var WIKI_PREFIX_URL =  'https://en.wikipedia.org/w/api.php?format=json&action=query&generator=search&gsrnamespace=0&gsrlimit=4&prop=pageimages|extracts&pilimit=max&exintro&explaintext&exsentences=1&exlimit=max&gsrsearch=';
+
+module.exports = function (callback,username,input){
+  'use strict';
+  let apiOptions = {
+    url :`${WIKI_PREFIX_URL}${input}`,
```
Comment (526avijitgupta) [inline services/wikiService.js]: space after `:`

**Comment by priyansh1161** [comment]
@526avijitgupta hope i have covered all the issues this time :octocat:

**Comment by 526avijitgupta** [comment]
@priyansh1161 Great work 👍 .. merging now.

**Pull request merged**